lib/org/chromium/webidl/Canonicalizer.js

```diff
@@ -0,0 +1,129 @@
+    must be done through one canonicalizer instances, or you will
+    get fragmented results.`,
+
+  requires: [
+    'org.chromium.webidl.ast.Enum',
+    'org.chromium.webidl.ast.Typedef',
+    'org.chromium.webidl.PipelineMessage',
+  ],
+
+  properties: [
+    {
+      class: 'Function',
+      documentation: `Function is called after waitTime seconds has passed
+        since the last call to addFragment.`,
+      name: 'onDone',
+      required: true,
+    },
+    {
+      class: 'Int',
+      documentation: `The amount of time to wait in seconds after the last
+        addFragment request, before onDone() is called with the canonical
+        IDL files.`,
+      name: 'waitTime',
+    },
+    {
+      documentation: `A map of maps containing canonical IDL files. The
+        key of the map corresponds to the renderer (source) from which the
+        files were extract. For example, rendererMap['blink'] is a map that
+        contains the canonical IDL files from the Blink repo.`,
+      name: 'rendererMap_',
+      factory: function() { return {}; },
+    },
+    {
+      documentation: `A map of Timers that is used to determine whether waitTime
+        has passed since the last call to addFragment.`,
+      name: 'timerMap_',
+      factory: function() { return {}; },
+    },
+  ],
+
+  methods: [
+    {
+      documentation: `Takes a given AST and extracts the definition members
+        from the AST. The members are then placed into the map corresponding to
+        the given source and interface.`,
+      name: 'addFragment',
+      code: function(source, ast, metadata) {
+        // Determine if Map exists for this source. Create one if non-existent.
+        var map = this.rendererMap_[source];
+        if (!map) {
+          map = {};
+          this.rendererMap_[source] = map;
+        }
+
+        var def = ast.definition;
+        // Ignore definition if Enum or Typedef.
+        if (this.Enum.isInstance(def) || this.Typedef.isInstance(def)) return;
+
+        // Find any existing interface with same id (name).
+        var clone = ast.clone();
+        var result = map[ast.id];
+
+        if (!result) {
+          // No interface with the same name was found.
+          // We add a copy of ourselves into the map.
+          clone.ref = [metadata];
+          map[ast.id] = clone;
+        } else {
+          // An interface with the same name was found.
+          var isCanonicalPartial = result.definition.isPartial;
+          var isDefPartial = def.isPartial;
+
+          // Determine if we have multiple non-partial fragments.
+          if (!isCanonicalPartial && !isDefPartial) {
+            console.error(`Two non-partial fragments!
+              In Map: ${result.ref},
+              Current: ${metadata}`);
+            throw new Error('Two non-partial fragments were found for the same interface!');
+          }
+
+          // Copy attributes and members from definition into canonical AST.
```

Comment:
To be clear, you would like metadata on where each member and attribute came from? It shouldn't be too hard to do, but it may make the code uglier (at least I think it will)